File: src/main/java/com/kasured/peuler/p001/Problem001.scala
object Problem001 {

    //boilerplate
    def main(args: Array[String]) {
        if(args == null || args.length < 1) {println(0)} else {
            val N = java.lang.Integer.parseInt(args(0), 10)
            println(solve(N))
        }
    }

    def solve(N: Long): Long = {
        //check edge case and go into tailrec dance
        if(N < 3) 0 else doSolve(N, 3, 1, 0) + doSolve(N, 5, 1, 0) - doSolve(N, 15, 1, 0)
    }

    @scala.annotation.tailrec
    def doSolve(N: Long, mod: Long, idx: Long, acc: Long): Long = {
        //calc current value
        val n = mod * idx;
        //check if it out of upper bound return accumulator
        //else do tailrec call with updated sum and increased index
        if(n >= N) acc else doSolve(N, mod, idx + 1, acc + n)
    }

}

File: src/main/java/com/kasured/peuler/p001/Problem001.java
class Problem001 {


    public static void main(final String[] args) {
        final int N = Integer.parseInt(args[0], 10);
        System.out.println(solve(N));
    }

    /**
     * @param N - the upper limit 
     */
    public static long solve(final long N) {
        if(N < 3) {
            return 0;
        }

        return doSolve(N, 3L, 1L, 0L) + doSolve(N, 5L, 1L, 0L) - doSolve(N, 15L, 1L, 0L);
    }

    @tailrec
    private static long doSolve(final long N, final long mod, final long idx, final long acc) {
        final long n = mod * idx;
        if(n >= N) {
            return acc;
        } else {
            return doSolve(N, mod, idx + 1, acc + n);
        }
    }

}

